### DISCUSSION TITLE
[Feature Request] Progress indicators for iterators with unknown length

### DISCUSSION THREAD

**Discussion by kescobo** [FedeClaudi/Term.jl#44] [Ideas]
I'm really enjoying getting to know Term.jl, and the docs are great! One thing that I'm missing is a progress bar for an iterator or loop with unknown length. Clearly, you could not have a percent or fraction, but something like a spinning ball + a counter for the number of iterations that have passed or time elapse would be really useful.

You can see [`ProgressMeter.jl`](https://github.com/timholy/ProgressMeter.jl#progress-meters-for-tasks-with-an-unknown-number-of-steps) for inspiration.

```
julia> mutable struct MyItr
           count
       end

julia> Base.iterate(mi::MyItr, state=mi.count) = state == 0 ? nothing : (state, state-1)

julia> let prog = ProgressUnknown("Here we go...", spinner=true)
       for i in MyItr(100)
           ProgressMeter.next!(prog)
           sleep(0.1)
       end
       ProgressMeter.finish!(prog)
       end
✓ Here we go...          Time: 0:00:10
```

**Comment by FedeClaudi** [comment]
Thanks for opening the first discussion!

This would be very good indeed, added to the list. 

You can almost get something like this done with the current code if you need it urgently"
```Julia
# create custom columns layout
import Term.progress: DescriptionColumn, ElapsedColumn, AbstractColumn, ProgressBar, update, start, stop
import Term

# define a new column type
struct CountsColumn <: AbstractColumn
    segments::Vector
    measure::Measure
    style::String
end

# constructor
CountsColumn(style::String) = CountsColumn([], Measure(10, 1), style)

# update method
update(col::CountsColumn, i::Int, args...)::String = apply_style("[$(col.style)]$i iterations[/$(col.style)]")


columns = [
    DescriptionColumn("[green bold]Job running[/green bold]"),
    CountsColumn("blue"),
    ElapsedColumn(),
]


# create iterable
mutable struct MyItr
    count
end

Base.iterate(mi::MyItr, state=mi.count) = state == 0 ? nothing : (state, state-1)


# ----------------------------- run progress bar ----------------------------- #
pbar = ProgressBar(; N=1000, columns=columns)  # set to very high value so that it gets ignored
start(pbar)

for i in MyItr(100)
    update(pbar)
    sleep(0.1)
end
stop(pbar)
```
<img width="327" alt="image" src="https://user-images.githubusercontent.com/17436313/158961707-94b711ae-9dab-48c8-a237-a2d18dd547e1.png">


The problem is that `ProgressBar` still expects an `N` argument and it will add a `BarColumn` if one is not provided. This is very easy to change: just let `ProgressBar` accept `N=nothing` and don't add the `BarColumn` in case. 

I can add it soon, but feel free to send a PR if you'd like!

**Reply by kescobo** [reply]
No rush! I have not had a chance to dive into the internals (and likely won't have time for at least a bit), so a PR is probably not going to happen in the short term, but I'm already imagining a lot of uses for this package! I definitely hope to contribute in the future.

**Reply by FedeClaudi** [reply]
Awesome! Looking forward to that!

Also let me know what you end up using Term for!

**Comment by FedeClaudi** [comment]
Working on it, keep an eye here https://github.com/FedeClaudi/Term.jl/pull/62

**Comment by FedeClaudi** [comment]
It will come with: https://github.com/FedeClaudi/Term.jl/pull/62

![ezgif com-gif-maker(7)](https://user-images.githubusercontent.com/17436313/160479353-33d3fab4-278b-4d03-9ad2-1afded32ffdb.gif)

Usage:
```Julia
pbar = ProgressBar(; columns=:spinner)
```
to use these preset columns. If you don't pass `N` you can still use any set of columns you'd like, but it just won't show a `ProgressColumn` with the progress bar any longer. 

The spinner visualization is handled by a new `SpinnerColumn` and multiple types of spinners are available, as shown (also you can apply markup style to the spinner).

Hope you'll like it!

**Reply by kescobo** [reply]
Love 'bar' - reminds me of Kit from Knight rider :laughing:

**Reply by kescobo** [reply]
Or maybe cylons...

**Comment by nathanrboyer** [comment]
I'd like to see the spinner change to a check mark when it is done. I'm also wondering how you got your previous spinners to stop. My first spinner restarts when I start the second one.
```julia
function f()
    with(progressbar) do
        readjob = addjob!(progressbar, description = "Reading Input File")
        start!(readjob)
        sleep(5)
        stop!(readjob)

        transformjob = addjob!(progressbar, description = "Transforming Tables")
        start!(transformjob)
        sleep(5)
        stop!(transformjob)
    end
end
progressbar = ProgressBar(; columns=:minimal, columns_kwargs = Dict(:SpinnerColumn => Dict(:spinnertype => :circle)))
f()
```

**Reply by FedeClaudi** [reply]
In the example above I was creating entirely new progress bars, this was before the new API with `addjob!` and the rest.
It looks like it's a bug and the spinner doesn't stop when calling `stop!` as one would would expect. It will require a small change to the update method of the spinner column [here](https://github.com/FedeClaudi/Term.jl/blob/51eee9aa0b3e40580f897df27b331cf6be5f26aa/src/_progress.jl#L337). 

In the meanwhile you can remove the job to hide the spinner:
```Julia
function f()
  with(progressbar) do
      readjob = addjob!(progressbar, description = "Reading Input File")
      start!(readjob)
      sleep(1)
      stop!(readjob)
      removejob!(progressbar, readjob)

      println("done")
      sleep(2)

      transformjob = addjob!(progressbar, description = "Transforming Tables")
      start!(transformjob)
      sleep(2)
      stop!(transformjob)
  end
end
```

> I'd like to see the spinner change to a check mark when it is done. 

That's a neat idea, I can add that in.

**Comment by FedeClaudi** [comment]
From the next release, calling `stop!` on a spinner progress bar not only stops the spinner, but also turns it into a checkmark:

<img width="561" alt="image" src="https://user-images.githubusercontent.com/17436313/169151262-223ff96a-9394-4b92-b064-cda52afccf04.png">

Thanks for the suggestion @nathanrboyer

**Reply by nathanrboyer** [reply]
This is working great for me with the `sleep` test case. However, the spinner never shows up if I am doing actual processing like computing something large or generating a plot. It just jumps right to the checkmark when it is done. MWE:
```julia
function f(sleeptime)
    progressbar = ProgressBar(; columns=:minimal, columns_kwargs = Dict(:SpinnerColumn => Dict(:spinnertype => :circle)))
    with(progressbar) do
        readjob = addjob!(progressbar, description = "Reading Input File")
        start!(readjob)
        sleep(sleeptime)
        stop!(readjob)

        transformjob = addjob!(progressbar, description = "Processing Input")
        start!(transformjob)
        collect(1:10^8) .^ (10^9) # Doing real work
        stop!(transformjob)

        writejob = addjob!(progressbar, description = "Writing Output File")
        start!(writejob)
        sleep(sleeptime)
        stop!(writejob)
    end
end
```
![IMG_4565](https://user-images.githubusercontent.com/65452054/169605107-a870022c-9857-4afe-bb8d-efab9fba30b0.gif)

Also, does the progress bar have to jump down to the bottom of the terminal? Thanks!

**Reply by FedeClaudi** [reply]
It's something that came up before: https://github.com/FedeClaudi/Term.jl/issues/77#issuecomment-1112246701

I think I'm doing something wrong with how the progress bar display is updated in the background (by a separate process). I'm not very strong on multi-processing in Julia yet, but it's something that I aim to fix with the next major release.
Adding a `sleep` statement to your loops should fix that - not a very elegant solution but it should do for now I hope.

> Also, does the progress bar have to jump down to the bottom of the terminal? Thanks!

It does, because that allows for any output to `stdout` to be printed above the progress bar without breaking the progress bar itself.

**Reply by nathanrboyer** [reply]
Adding `sleep` to each section helped somewhat but didn't fix everything. Three of my spinners showed up but didn't spin. The last spinner (for plotting) never showed up. The spinners aren't critical I suppose, but it is nice to know what the code is working on.

**Reply by FedeClaudi** [reply]
I agree, I always add progress information or I get too impatient if I don't know what is being run... sorry that this is still a bit buggy!

**Reply by FedeClaudi** [reply]
This was fixed thanks to the brilliant folks at [Julia discourse](https://discourse.julialang.org/t/printing-in-a-background-process/81453). Will come out with the next version `v0.4`.
